File: ruby/lib/spreen/wiki/application.rb
# frozen_string_literal: true
# rbs_inline: enabled

require_relative 'configuration'

module Spreen
  module Wiki
    # Base class for the wiki-organising commands. Loads the wiki tree under
    # base_path, groups it by Owner or Category in either English or Japanese,
    # and exposes the resulting maps to its subclasses. Grouping criteria,
    # languages and labels are resolved through Configuration and can be
    # extended via a `.spreen.yml` file.
    class Application
      class NotImplementedError < StandardError; end

      # @rbs base_path: String
      # @rbs group_by: String
      # @rbs language: String
      # @rbs home_overflow: (bool | String)
      # @rbs **options: untyped
      # @rbs return: untyped
      def self.run(base_path: Dir.pwd, group_by: 'Owner', language: 'English', home_overflow: 'false', **)
        instance = new(base_path:, group_by:, language:, home_overflow:, **)
        instance.validate!
        instance.run
      end

      # @rbs base_path: String
      # @rbs group_by: String
      # @rbs language: String
      # @rbs home_overflow: (bool | String)
      # @rbs config: Configuration?
      # @rbs **options: untyped
      # @rbs return: void
      def initialize(base_path: Dir.pwd, group_by: 'Owner', language: 'English', home_overflow: 'false',
                     config: nil, **)
        @base_path       = base_path
        @group_by        = group_by
        @language        = language
        @home_overflow   = parse_home_overflow(home_overflow)
        @config          = config || Configuration.new(base_path:, **)
        @path_to_home    = File.join(base_path, 'Home.md')
        @path_to_sidebar = File.join(base_path, '_Sidebar.md')
        @paths_to_wikis  = Dir[File.join(base_path, '**', '*.md')]
        @excluded_paths  = @config.excluded_dirs.flat_map { |dir| Dir[File.join(base_path, dir, '**', '*.md')] }
      end

      # @rbs return: void
      def validate!
        raise ArgumentError, "Invalid group_by: `#{group_by}`" unless config.group_bys.include?(group_by)
        raise ArgumentError, "Invalid language: `#{language}`" unless config.languages(group_by).include?(language)

        validate_home_overflow!
      end

      # @rbs return: untyped
      def run
        raise NotImplementedError, 'This method must be implemented in each subclass.'
      end

      private

      attr_reader :base_path,
                  :group_by,
                  :language,
                  :home_overflow,
                  :config,
                  :path_to_home,
                  :path_to_sidebar,
                  :paths_to_wikis,
                  :excluded_paths

      # @rbs return: void
      def validate_home_overflow!
        return if [true, false].include?(home_overflow)

        raise ArgumentError, "Invalid home_overflow: `#{home_overflow}` must be boolean"
      end

      # @rbs raw: (bool | String)
      # @rbs return: (bool | String)
      def parse_home_overflow(raw)
        case raw
        when 'true'  then true
        when 'false' then false
        else raw
        end
      end

      # @rbs return: Array[String]
      def target_paths
        @target_paths ||= paths_to_wikis.reject do |path_to_wiki|
          path_to_wiki == path_to_home ||
            path_to_wiki == path_to_sidebar ||
            excluded_paths.include?(path_to_wiki)
        end
      end

      # @rbs return: Regexp
      def target_regexp
        @target_regexp ||= config.target_regexp(group_by)
      end

      # @rbs return: String
      def no_declaration
        @no_declaration ||= config.no_declaration(group_by, language)
      end

      # @rbs return: Hash[String, Array[String]]
      def wiki_maps_with_namespace
        @wiki_maps_with_namespace ||= begin
          named      = Hash.new { |hash, namespace| hash[namespace] = [] } #: Hash[String, Array[String]]
          undeclared = Hash.new { |hash, namespace| hash[namespace] = [] } #: Hash[String, Array[String]]
          target_paths.each { |target_path| populate_namespace(target_path, named, undeclared) }
          named.sort.to_h.merge(undeclared)
        end
      end

      # @rbs target_path: String
      # @rbs named: Hash[String, Array[String]]
      # @rbs undeclared: Hash[String, Array[String]]
      # @rbs return: void
      def populate_namespace(target_path, named, undeclared)
        return unless File.exist?(target_path)

        namespace = namespace_for(target_path)
        bucket    = namespace == no_declaration ? undeclared : named
        bucket[namespace] << File.basename(target_path)
      end

      # @rbs target_path: String
      # @rbs return: String
      def namespace_for(target_path)
        declaration = File.open(target_path, &:gets)
        return no_declaration unless declaration&.start_with?(target_regexp)

        declaration.chomp.gsub(target_regexp, '')
      end

      # @rbs return: Hash[String, Array[String]]
      def owned_wiki_maps
        @owned_wiki_maps ||= wiki_maps_with_namespace.select { |namespace, _| namespace.include?('@') }
      end

      # @rbs return: Hash[String, Array[String]]
      def plain_wiki_maps
        @plain_wiki_maps ||= wiki_maps_with_namespace.reject { |namespace, _| namespace.include?('@') }
      end
    end
  end
end

File: PyPI/src/spreen_wiki/application.py
import glob
import os
import re
from typing import Any

from .configuration import Configuration


class Application:
    """Base class for the wiki-organising commands. Loads the wiki tree
    under base_path, groups it by Owner or Category in either English or
    Japanese, and exposes the resulting maps to its subclasses. Grouping
    criteria, languages and labels are resolved through Configuration and
    can be extended via a `.spreen.yml` file."""

    class NotImplementedError(Exception):
        pass

    _cached_wiki_maps: dict[str, list[str]]

    @classmethod
    def run(cls, base_path: str | None = None, group_by: str = 'Owner',
            language: str = 'English',
            home_overflow: str | bool = 'false',
            **options: Any) -> Any:
        instance = cls(
            base_path=base_path, group_by=group_by,
            language=language, home_overflow=home_overflow, **options,
        )
        instance.validate()
        return instance._run()

    def __init__(self, base_path: str | None = None,
                 group_by: str = 'Owner', language: str = 'English',
                 home_overflow: str | bool = 'false',
                 config: Configuration | None = None,
                 **options: Any) -> None:
        self._base_path = base_path if base_path is not None else os.getcwd()
        self._group_by = group_by
        self._language = language
        self._home_overflow = self._parse_home_overflow(home_overflow)
        self._config = config or Configuration(
            base_path=self._base_path, **options)
        self._path_to_home = os.path.join(self._base_path, 'Home.md')
        self._path_to_sidebar = os.path.join(self._base_path, '_Sidebar.md')
        self._paths_to_wikis = sorted(glob.glob(
            os.path.join(self._base_path, '**', '*.md'), recursive=True))
        self._excluded_paths = {
            path
            for excluded_dir in self._config.excluded_dirs
            for path in glob.glob(
                os.path.join(self._base_path, excluded_dir, '**', '*.md'),
                recursive=True)
        }

    def validate(self) -> None:
        if self._group_by not in self._config.group_bys():
            raise ValueError(f'Invalid group_by: `{self._group_by}`')
        if self._language not in self._config.languages(self._group_by):
            raise ValueError(f'Invalid language: `{self._language}`')
        if self._home_overflow not in (True, False):
            raise ValueError(
                f'Invalid home_overflow: `{self._home_overflow}` '
                'must be boolean')

    # private

    def _run(self) -> Any:
        raise self.NotImplementedError(
            'This method must be implemented in each subclass.')

    @staticmethod
    def _parse_home_overflow(raw: str | bool) -> str | bool:
        match raw:
            case 'true':
                return True
            case 'false':
                return False
            case _:
                return raw

    def _target_paths(self) -> list[str]:
        excluded = {self._path_to_home, self._path_to_sidebar}
        excluded.update(self._excluded_paths)
        return [p for p in self._paths_to_wikis if p not in excluded]

    def _target_regexp(self) -> re.Pattern[str]:
        return self._config.target_regexp(self._group_by)

    def _no_declaration(self) -> str:
        return self._config.no_declaration(self._group_by, self._language)

    def _wiki_maps_with_namespace(self) -> dict[str, list[str]]:
        if hasattr(self, '_cached_wiki_maps'):
            return self._cached_wiki_maps
        named: dict[str, list[str]] = {}
        undeclared: dict[str, list[str]] = {}
        for target_path in self._target_paths():
            self._populate_namespace(target_path, named, undeclared)
        result = dict(sorted(named.items()))
        result.update(undeclared)
        self._cached_wiki_maps = result
        return result

    def _populate_namespace(self, target_path: str,
                            named: dict[str, list[str]],
                            undeclared: dict[str, list[str]]) -> None:
        if not os.path.isfile(target_path):
            return
        with open(target_path, encoding='utf-8') as f:
            wiki = os.path.basename(target_path)
            namespace = self._namespace_for(f)
        target = undeclared if namespace == self._no_declaration() \
            else named
        target.setdefault(namespace, []).append(wiki)

    def _namespace_for(self, file: Any) -> str:
        declaration = file.readline()
        if not declaration or not self._target_regexp().match(declaration):
            return self._no_declaration()
        return re.sub(self._target_regexp(), '', declaration.rstrip('\n'))

    def _owned_wiki_maps(self) -> dict[str, list[str]]:
        return {
            namespace: wikis
            for namespace, wikis in self._wiki_maps_with_namespace().items()
            if '@' in namespace
        }

    def _plain_wiki_maps(self) -> dict[str, list[str]]:
        return {
            namespace: wikis
            for namespace, wikis in self._wiki_maps_with_namespace().items()
            if '@' not in namespace
        }
